### PR TITLE
Add (stateful) consecutive predicate and update object settled predicate

### PR DESCRIPTION
**Pull request by xyao-nv**
## Summary
Add (stateful) consecutive predicate and update object settled predicate

## Detailed description
- Added `ConsecutivePredicate` and `ObjectsSettledForConsecutiveSteps` with reset-aware lifecycle management.
- Extended progress tracking to support callable predicates and managed `TerminationTermCfg` predicates.
- Added ObjectsSettledTask into the reusable task library as an example to show the object settled task

## Example: object settled for 10 frames as the only task predicate
https://github.com/user-attachments/assets/547bbed5-6fe4-4e49-a8c4-37ab8cedb96b

## TODO
PredicateGroup will be added when there are more than 1 predicate in the task, i.e. gear insertion in the next MR.

### CODE REVIEW

File: isaaclab_arena/tasks/predicates/consecutive.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# Copyright (c) 2026, The Isaac Lab Arena Project Developers (https://github.com/isaac-sim/IsaacLab-Arena/blob/main/CONTRIBUTORS.md).
```
Comment (peterd-NV) [inline isaaclab_arena/tasks/predicates/consecutive.py:1]: Is there a reason why this file is called `composition.py`? The naming seems a bit confusing since this file contains the `ConsecutivePredicate` class.

File: isaaclab_arena/tasks/predicates/object_settling.py [hunk 2]
```diff
@@ -117,14 +150,60 @@ def objects_settled(
...
+        assert all(
+            isinstance(name, str) and name for name in object_names
+        ), f"ObjectsSettledForConsecutiveSteps object names must be non-empty strings, got {object_names!r}."
+        assert (
+            lin_vel_threshold >= 0.0
+        ), f"ObjectsSettledForConsecutiveSteps linear velocity threshold must be non-negative, got {lin_vel_threshold}."
+        assert ang_vel_threshold >= 0.0, (
+            "ObjectsSettledForConsecutiveSteps angular velocity threshold must be non-negative, got"
+            f" {ang_vel_threshold}."
+        )
```
[the remark this answers is not in the document]
Comment (peterd-NV) [reply isaaclab_arena/tasks/predicates/object_settling.py]: This seems like a reasonable suggestion.

File: isaaclab_arena/tasks/predicates/composition.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    def _update_consecutive(self, passed: torch.Tensor) -> torch.Tensor:
+        """Update per-environment streaks and return which have reached the required length."""
+
+        passed = torch.as_tensor(passed, dtype=torch.bool, device=self.device).reshape(-1)
+        assert passed.shape == (
+            self.num_envs,
+        ), f"Predicate returned shape {tuple(passed.shape)}; expected ({self.num_envs},)."
+        next_count = torch.clamp(self.consecutive_true_steps + 1, max=self._required_consecutive_steps)
+        self.consecutive_true_steps = torch.where(passed, next_count, torch.zeros_like(self.consecutive_true_steps))
```
[the remark this answers is not in the document]
Comment (xyao-nv) [reply isaaclab_arena/tasks/predicates/composition.py]: added an active mask, shall be fixed

File: isaaclab_arena/tasks/predicates/object_settling.py [hunk 4]
```diff
@@ -117,14 +150,60 @@ def objects_settled(
...
+        recorder = get_rest_pose_recorder(env)
+        for object_name in object_names:
+            recorder.record(object_name, env.arena_world.get_position_w(object_name), settled)
+        return settled
+
+    def reset(self, env_ids: Sequence[int] | None = None) -> None:
+        """Clear stability counters and recorded rest poses for selected environments."""
+
+        super().reset(env_ids)
+        get_rest_pose_recorder(self._env).reset(env_ids)
```
[the remark this answers is not in the document]
Comment (xyao-nv) [reply isaaclab_arena/tasks/predicates/object_settling.py]: removed the record in progress track reset

File: isaaclab_arena/progress_tracking/progress_tracker.py [hunk 5]
```diff
@@ -20,6 +22,23 @@
...
 
 
+def _resolve_progress_predicate(predicate, env):
+    """Instantiate a managed predicate config when the tracker gains access to the environment."""
+
+    if not isinstance(predicate, TerminationTermCfg):
+        return predicate
+
+    assert env is not None, "An environment is required to resolve a managed progress predicate."
+    predicate_cfg = copy.deepcopy(predicate)
```
[the remark this answers is not in the document]
Comment (xyao-nv) [reply isaaclab_arena/progress_tracking/progress_tracker.py:32]: deeop copy only for predictaes owned exclusively by progress tracking

File: isaaclab_arena/tasks/predicates/object_settling.py [hunk 6]
```diff
@@ -117,14 +150,60 @@ def objects_settled(
...
+        ), f"ObjectsSettledForConsecutiveSteps object names must be non-empty strings, got {object_names!r}."
+        assert (
+            lin_vel_threshold >= 0.0
+        ), f"ObjectsSettledForConsecutiveSteps linear velocity threshold must be non-negative, got {lin_vel_threshold}."
+        assert ang_vel_threshold >= 0.0, (
+            "ObjectsSettledForConsecutiveSteps angular velocity threshold must be non-negative, got"
+            f" {ang_vel_threshold}."
+        )
+
+    def __call__(
```
Comment (peterd-NV) [inline isaaclab_arena/tasks/predicates/object_settling.py:187]: I think there is a hidden bug here when this is used with the progress tracker and parallel envs. The progress tracker evaluates predicates via something like result = predicate(env). However, when running parallel envs, not all envs may have the same active predicates. This is handled after the predicate is evaluated using a gating mask. 

This is not an issue for state less predicates as there is no internal counter/state. However, for a stateful predicate, this means that an env could be accumulating "counts" before the predicate is even active. 

I think we need an optional additional argument here called `active_mask` that gates the internal state of the predicate from updating for envs where the predicate is not just active. Something like: 

`result = predicate(env, active_mask=at_position)`

See the comment above in `progress_tracker.py` where I flag where this call to the predicate is happening.

File: isaaclab_arena/tasks/predicates/object_settling.py [hunk 7]
```diff
@@ -102,6 +112,29 @@ def objects_settled(
     respective thresholds. The recorded rest poses are readable via ``get_object_initial_rest_state``.
     """
 
+    settled = _objects_below_velocity_thresholds(
```
[the remark this answers is not in the document]
Comment (xyao-nv) [reply isaaclab_arena/tasks/predicates/object_settling.py:123]: bring back objects_settled

File: isaaclab_arena/progress_tracking/progress_tracker.py
Comment (peterd-NV) [inline isaaclab_arena/progress_tracking/progress_tracker.py]: Flagging this line to show where the predicate is evaluated: The predicate is called here for all envs. For a stateful predicate, this means that the internal state could be incrementing even if the predicate is not yet active in the chain yet.

File: isaaclab_arena/progress_tracking/progress_tracker.py (hunk 5 above)
Comment (peterd-NV) [inline isaaclab_arena/progress_tracking/progress_tracker.py:32]: Note: I think this means that the progress tracker is using a separate predicate object as the termination term. Each env step will require the same predicate running twice and updating. Could lead to unexpected divergences between progress tracking and terminations. 

Assuming we unify progress tracking and success terminations, then I don't think this will be an issue. But just something to keep in mind.

**Comment by xyao-nv** [comment]
@arena-review-bot review

**Review by peterd-NV — COMMENTED** [review]
Looks good overall. One major comment relating to correctness of stateful predicates for multi-env case that I think needs to be addressed.

File: isaaclab_arena/tasks/predicates/object_settling.py (hunk 6 above)
Comment (peterd-NV) [reply isaaclab_arena/tasks/predicates/object_settling.py:187]: This has been resolved.

**Review by peterd-NV — APPROVED** [review]
Thanks for fixing the mult-env issue. LGTM. Other comments are minor nits.

File: isaaclab_arena/progress_tracking/progress_tracker.py (hunk 5 above)
Comment (xyao-nv) [reply isaaclab_arena/progress_tracking/progress_tracker.py:32]: The deepcopy is only for those predicates not used in the task, but added in the progress tracking.

File: isaaclab_arena/tasks/predicates/consecutive.py (hunk 1 above)
Comment (xyao-nv) [reply isaaclab_arena/tasks/predicates/consecutive.py:1]: renamed

File: isaaclab_arena/progress_tracking/progress_tracker.py [hunk 8]
```diff
@@ -434,7 +484,6 @@ def progress_tracking_reset_func(env, env_ids, progress_objectives: list[Progres
     elif torch.is_tensor(env_ids):
         env_ids = env_ids.tolist()
     progress_tracker.reset(env_ids)
-    reset_rest_pose_recorder(env, env_ids)
```
Comment (cvolkcvolk) [inline isaaclab_arena/progress_tracking/progress_tracker.py:437]: This affects existing progress tracking in PickAndPlace no? That works for the new managed predicate because `ObjectsSettledForConsecutiveSteps.reset() ` clears its recorded positions. However, `PickAndPlaceTask` still uses the old, stateless `objects_settled` _function_. That function has no reset method, and the environment doesn’t clear the recorder elsewhere?

File: isaaclab_arena/tests/objects_settled_task.py
Comment (cvolkcvolk) [inline isaaclab_arena/tests/objects_settled_task.py:1]: Is this Taskl intended as a reusable task, or mainly as an integration example and testing? Suggestion to move it to the test/ directory

File: isaaclab_arena/tasks/terminations.py [hunk 9]
```diff
@@ -32,6 +32,12 @@ class SuccessMode(str, Enum):
     """Success needs at least k predicates to be True."""
 
 
+def termination_term_result(env: ManagerBasedRLEnv, term_name: str) -> torch.Tensor:
```
Comment (cvolkcvolk) [inline isaaclab_arena/tasks/terminations.py:37]: Suggestion to add

`TODO(cvolk): Remove this success-to-progress adapter once its callers use tracker-owned stateful predicates through TaskTerminationCfg `

File: isaaclab_arena/progress_tracking/progress_tracker.py [hunk 10]
```diff
@@ -5,21 +5,49 @@
...
 
 from isaaclab_arena.progress_tracking.progress_objective import ProgressObjective, ProgressObjectiveCompletionMode
 from isaaclab_arena.progress_tracking.progress_tracking_utils import _predicate_repr
-from isaaclab_arena.tasks.predicates.object_settling import reset_rest_pose_recorder
+from isaaclab_arena.tasks.predicates.consecutive import ConsecutivePredicate
 
 _PROGRESS_TRACKER_ATTR = "_progress_tracker"
 
 
+def _resolve_progress_predicate(predicate, env):
```
Comment (cvolkcvolk) [inline isaaclab_arena/progress_tracking/progress_tracker.py:26]: Suggestion to add a note for me to come back to this

`TODO(cvolk): This adapts Isaac Lab term configs for progress tracking. Revisit the adapter when integrating stateful predicates with #1255. predicate construction and reset forwarding must remain supported.`

**Review by cvolkcvolk — APPROVED** [review]
Lgtm. One comment about implications on the current PickAndPlace task

File: isaaclab_arena/progress_tracking/progress_tracker.py (hunk 8 above)
Comment (xyao-nv) [reply isaaclab_arena/progress_tracking/progress_tracker.py:437]: hm good catch. I wasn't aware of that. Added

File: isaaclab_arena/progress_tracking/progress_tracker.py (hunk 10 above)
Comment (xyao-nv) [reply isaaclab_arena/progress_tracking/progress_tracker.py:26]: YESS
